File: example_proj/routes/teams.py
from flask import Blueprint, jsonify, request
from flask_login import login_required, current_user
from models.user import User
from app import db, limiter
from datetime import datetime

teams_bp = Blueprint('teams', __name__)

@teams_bp.route('/teams', methods=['GET'])
@login_required
@limiter.limit('30 per minute')
def get_teams():
    """Get all teams for the current user."""
    try:
        teams = current_user.teams
        return jsonify([{
            'id': team.get('id'),
            'name': team.get('name'),
            'role': team.get('role'),
            'created_at': team.get('created_at')
        } for team in teams])
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@teams_bp.route('/teams', methods=['POST'])
@login_required
@limiter.limit('10 per minute')
def create_team():
    """Create a new team."""
    try:
        data = request.get_json()
        team_name = data.get('name')
        
        if not team_name:
            return jsonify({'error': 'Team name is required'}), 400
            
        # Check if user can create more teams based on subscription
        team_limit = current_user.subscription.get('team_limit', 1)
        if len(current_user.teams) >= team_limit:
            return jsonify({'error': 'Team limit reached for your subscription'}), 403
        
        # Create new team
        new_team = {
            'id': str(len(current_user.teams) + 1),  # Simple ID generation
            'name': team_name,
            'role': 'owner',
            'created_at': datetime.utcnow().isoformat(),
            'members': [{
                'user_id': current_user.id,
                'role': 'owner',
                'joined_at': datetime.utcnow().isoformat()
            }]
        }
        
        current_user.teams.append(new_team)
        db.session.commit()
        
        return jsonify(new_team), 201
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@teams_bp.route('/teams/<team_id>', methods=['PUT'])
@login_required
@limiter.limit('10 per minute')
def update_team(team_id):
    """Update team details."""
    try:
        data = request.get_json()
        team_name = data.get('name')
        
        # Find team
        team = next((t for t in current_user.teams if t['id'] == team_id), None)
        if not team:
            return jsonify({'error': 'Team not found'}), 404
            
        # Check permissions
        if team['role'] not in ['owner', 'admin']:
            return jsonify({'error': 'Unauthorized'}), 403
        
        # Update team
        team['name'] = team_name
        db.session.commit()
        
        return jsonify(team)
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@teams_bp.route('/teams/<team_id>/members', methods=['POST'])
@login_required
@limiter.limit('10 per minute')
def add_team_member(team_id):
    """Add a new member to the team."""
    try:
        data = request.get_json()
        email = data.get('email')
        role = data.get('role', 'member')
        
        if not email:
            return jsonify({'error': 'Email is required'}), 400
            
        # Find team
        team = next((t for t in current_user.teams if t['id'] == team_id), None)
        if not team:
            return jsonify({'error': 'Team not found'}), 404
            
        # Check permissions
        if team['role'] not in ['owner', 'admin']:
            return jsonify({'error': 'Unauthorized'}), 403
        
        # Find user by email
        user = User.query.filter_by(email=email).first()
        if not user:
            return jsonify({'error': 'User not found'}), 404
            
        # Check if user is already a member
        if any(m['user_id'] == user.id for m in team['members']):
            return jsonify({'error': 'User is already a team member'}), 400
        
        # Add member
        team['members'].append({
            'user_id': user.id,
            'role': role,
            'joined_at': datetime.utcnow().isoformat()
        })
        
        # Add team to user's teams
        user_team = team.copy()
        user_team['role'] = role
        user.teams.append(user_team)
        
        db.session.commit()
        
        return jsonify({'message': 'Member added successfully'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@teams_bp.route('/teams/<team_id>/members/<user_id>', methods=['DELETE'])
@login_required
@limiter.limit('10 per minute')
def remove_team_member(team_id, user_id):
    """Remove a member from the team."""
    try:
        # Find team
        team = next((t for t in current_user.teams if t['id'] == team_id), None)
        if not team:
            return jsonify({'error': 'Team not found'}), 404
            
        # Check permissions
        if team['role'] not in ['owner', 'admin']:
            return jsonify({'error': 'Unauthorized'}), 403
        
        # Remove member
        team['members'] = [m for m in team['members'] if m['user_id'] != int(user_id)]
        
        # Remove team from user's teams
        user = User.query.get(user_id)
        if user:
            user.teams = [t for t in user.teams if t['id'] != team_id]
        
        db.session.commit()
        
        return jsonify({'message': 'Member removed successfully'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@teams_bp.route('/teams/<team_id>', methods=['DELETE'])
@login_required
@limiter.limit('10 per minute')
def delete_team(team_id):
    """Delete a team."""
    try:
        # Find team
        team = next((t for t in current_user.teams if t['id'] == team_id), None)
        if not team:
            return jsonify({'error': 'Team not found'}), 404
            
        # Check permissions
        if team['role'] != 'owner':
            return jsonify({'error': 'Unauthorized'}), 403
        
        # Remove team from all members
        for member in team['members']:
            user = User.query.get(member['user_id'])
            if user:
                user.teams = [t for t in user.teams if t['id'] != team_id]
        
        # Remove team from current user
        current_user.teams = [t for t in current_user.teams if t['id'] != team_id]
        
        db.session.commit()
        
        return jsonify({'message': 'Team deleted successfully'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

File: example_proj/server/routes/teams.js
import express from 'express';
import { getService } from '../services/index.js';
import { validateRequest } from '../middleware/validation.js';

const router = express.Router();

// Get user's teams
router.get('/', async (req, res) => {
  try {
    const analyticsService = getService('analytics');
    const userId = req.user.id;
    
    const teams = await analyticsService.getUserTeams(userId);
    
    res.json({
      success: true,
      data: teams
    });
  } catch (error) {
    console.error('Get teams error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to fetch teams',
      error: error.message
    });
  }
});

// Create a new team
router.post('/', async (req, res) => {
  try {
    const analyticsService = getService('analytics');
    const userId = req.user.id;
    const { name, description, settings } = req.body;
    
    const team = await analyticsService.createTeam(userId, {
      name,
      description,
      settings
    });
    
    res.status(201).json({
      success: true,
      data: team
    });
  } catch (error) {
    console.error('Create team error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to create team',
      error: error.message
    });
  }
});

// Get team details
router.get('/:teamId', async (req, res) => {
  try {
    const analyticsService = getService('analytics');
    const userId = req.user.id;
    const { teamId } = req.params;
    
    const team = await analyticsService.getTeamDetails(userId, teamId);
    
    res.json({
      success: true,
      data: team
    });
  } catch (error) {
    console.error('Get team details error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to fetch team details',
      error: error.message
    });
  }
});

// Update team
router.put('/:teamId', async (req, res) => {
  try {
    const analyticsService = getService('analytics');
    const userId = req.user.id;
    const { teamId } = req.params;
    const updateData = req.body;
    
    const team = await analyticsService.updateTeam(userId, teamId, updateData);
    
    res.json({
      success: true,
      data: team
    });
  } catch (error) {
    console.error('Update team error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to update team',
      error: error.message
    });
  }
});

// Delete team
router.delete('/:teamId', async (req, res) => {
  try {
    const analyticsService = getService('analytics');
    const userId = req.user.id;
    const { teamId } = req.params;
    
    await analyticsService.deleteTeam(userId, teamId);
    
    res.json({
      success: true,
      message: 'Team deleted successfully'
    });
  } catch (error) {
    console.error('Delete team error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to delete team',
      error: error.message
    });
  }
});

// Add member to team
router.post('/:teamId/members', async (req, res) => {
  try {
    const analyticsService = getService('analytics');
    const userId = req.user.id;
    const { teamId } = req.params;
    const { email, role = 'member' } = req.body;
    
    const member = await analyticsService.addTeamMember(userId, teamId, {
      email,
      role
    });
    
    res.status(201).json({
      success: true,
      data: member
    });
  } catch (error) {
    console.error('Add team member error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to add team member',
      error: error.message
    });
  }
});

// Remove member from team
router.delete('/:teamId/members/:memberId', async (req, res) => {
  try {
    const analyticsService = getService('analytics');
    const userId = req.user.id;
    const { teamId, memberId } = req.params;
    
    await analyticsService.removeTeamMember(userId, teamId, memberId);
    
    res.json({
      success: true,
      message: 'Team member removed successfully'
    });
  } catch (error) {
    console.error('Remove team member error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to remove team member',
      error: error.message
    });
  }
});

// Update member role
router.put('/:teamId/members/:memberId', async (req, res) => {
  try {
    const analyticsService = getService('analytics');
    const userId = req.user.id;
    const { teamId, memberId } = req.params;
    const { role } = req.body;
    
    const member = await analyticsService.updateMemberRole(userId, teamId, memberId, role);
    
    res.json({
      success: true,
      data: member
    });
  } catch (error) {
    console.error('Update member role error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to update member role',
      error: error.message
    });
  }
});

// Get team documents
router.get('/:teamId/documents', async (req, res) => {
  try {
    const analyticsService = getService('analytics');
    const userId = req.user.id;
    const { teamId } = req.params;
    const { page = 1, limit = 20, search } = req.query;
    
    const documents = await analyticsService.getTeamDocuments(userId, teamId, {
      page: parseInt(page),
      limit: parseInt(limit),
      search
    });
    
    res.json({
      success: true,
      data: documents
    });
  } catch (error) {
    console.error('Get team documents error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to fetch team documents',
      error: error.message
    });
  }
});

// Share document with team
router.post('/:teamId/documents/:documentId/share', async (req, res) => {
  try {
    const analyticsService = getService('analytics');
    const userId = req.user.id;
    const { teamId, documentId } = req.params;
    const { permissions = 'read' } = req.body;
    
    const shareResult = await analyticsService.shareDocumentWithTeam(userId, teamId, documentId, permissions);
    
    res.json({
      success: true,
      data: shareResult
    });
  } catch (error) {
    console.error('Share document with team error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to share document with team',
      error: error.message
    });
  }
});

export default router;